cpp/src/table/row_operators.cu

```diff
@@ -0,0 +1,203 @@
+          curr_col = prev_col;
+        } else {
+          curr_col = column_view(prev_col.type(),
+                                 prev_col.size(),
+                                 nullptr,
+                                 prev_col.null_mask(),
+                                 UNKNOWN_NULL_COUNT,
+                                 prev_col.offset(),
+                                 {curr_col});
+        }
+        --curr_col_idx;
+      }
+      verticalized_columns.insert(
+        verticalized_columns.end(), r_verticalized_columns.rbegin(), r_verticalized_columns.rend());
+      verticalized_col_depths.insert(verticalized_col_depths.end(),
+                                     r_verticalized_col_depths.rbegin(),
+                                     r_verticalized_col_depths.rend());
+      if (not column_order.empty()) {
+        new_column_order.insert(
+          new_column_order.end(), r_verticalized_columns.size(), column_order[col_idx]);
+      }
+      if (not null_precedence.empty()) {
+        new_null_precedence.insert(
+          new_null_precedence.end(), r_verticalized_columns.size(), null_precedence[col_idx]);
+      }
+    } else {
+      verticalized_columns.push_back(col);
+      verticalized_col_depths.push_back(0);
+      if (not column_order.empty()) { new_column_order.push_back(column_order[col_idx]); }
+      if (not null_precedence.empty()) { new_null_precedence.push_back(null_precedence[col_idx]); }
+    }
+  }
+  return std::make_tuple(table_view(verticalized_columns),
+                         std::move(new_column_order),
+                         std::move(new_null_precedence),
+                         std::move(verticalized_col_depths));
+}
+
+struct is_relationally_comparable_functor {
+  template <typename T>
+  constexpr bool operator()()
+  {
+    return cudf::is_relationally_comparable<T, T>();
+  }
+};
+
+/**
+ * @brief Check a table for compatibility with lexicographic comparison
+ *
+ * Checks whether a given table contains columns of non-relationally comparable types.
+ */
+void check_lex_compatibility(table_view const& input)
+{
+  // Basically check if there's any LIST hiding anywhere in the table
+  std::function<void(column_view const&)> check_column = [&](column_view const& c) {
+    CUDF_EXPECTS(c.type().id() != type_id::LIST,
+                 "Cannot lexicographic compare a table with a LIST column");
+    if (not is_nested(c.type())) {
+      CUDF_EXPECTS(
+        type_dispatcher<non_nested_id_to_type>(c.type(), is_relationally_comparable_functor{}),
+        "Cannot lexicographic compare a table with a column of type " +
+          jit::get_type_name(c.type()));
+    }
+    for (auto child = c.child_begin(); child < c.child_end(); ++child) {
+      check_column(*child);
+    }
+  };
+  for (column_view const& c : input) {
+    check_column(c);
+  }
+}
+
+}  // namespace
+
+namespace lex {
+
+preprocessed_table::preprocessed_table(table_view const& t,
```

Comment:
It probably doesn't matter for performance, but just double-checking: do we need `self_comparator`'s comparator to be inlined, or just the call operator? I assume just the operator, so I think we're fine, but if it's important for the constructor to be inlined I think this function needs to be marked inline since it's not defined in the header.

Reply:
These are all host objects and inline is not important for their constructors. Do you mean `device_row_comparator`? Original `row_lexicographic_comparator`'s ctor is not inline so I didn't think it's important. Even when a similar comparator is [wrapped in another comparator](https://github.com/rapidsai/cudf/blob/496aa47973aa89daf5a3f0b0b76405a8c7f2b92e/cpp/src/groupby/sort/sort_helper.cu#L53-L86), the constructor is not invoked per-row or anything.